Sources/NIOCore/ByteBuffer-core.swift

```diff
@@ -1119,3 +1117,105 @@ extension ByteBuffer {
         return Range<Int>(uncheckedBounds: (lower: indexFromReaderIndex, upper: upperBound))
     }
 }
+
+// MARK: HexDump helpers
+
+extension ByteBuffer {
+
+    /// Return a `String` of hexadecimal digits of bytes in the Buffer,
+    /// in a format that's compatible with `xxd -r -p`.
+    ///
+    /// - parameters:
+    ///     - offset:  The offset from the beginning of the buffer, in bytes. Defaults to 0.
+    ///     - readableOnly: Whether to only dump the `readableBytes` part (before the `writerIndex`), or to dump the full buffer, including empty bytes in the end of the allocated space. Defaults to true.
+    func hexDumpShort(offset: Int = 0, readableOnly: Bool = true) -> String {
+        let length = (readableOnly ? self.writerIndex : self.capacity) - offset
```

Comment:
The actual `dumpBytes` uses unsafe pointer and can dump any indices, so the dump will start at `self.bytes[offset]`.

`readableOnly` is an attempt to provide a convenience API to only dump the meaningful bytes that have been previously written, hence using `writerIndex`. 

Unlike `readableBytes`, even if the buffer was previously read from, and `readerIndex` moved, `writerIndex` is more likely to be useful.

Pros:
- Perhaps it's easier to dump the part of the buffer that was previously written to, less cleanup to do.

Cons:
- Naming is not yet final, and I think it does not make much sense. `readableOnly` implies dumping readable bytes, i.e. between `readerIndex..<writerIndex`, which is NOT what this is doing. 
- Unreliable. Someone could `setWriterIndex()` before calling this, and miss some parts of the buffer bytes that were useful to them.


I think if we keep this in the final version, this needs to be documented. Or we can remove that `readableOnly` option altogether, and just dump all buffer bytes to capacity, or accept `length` as an argument, default to full capacity. 

/cc @weissi, @glbrntt

Reply:
@natikgadzhi the `readableOnly` parameter needs to go. Nothing must _ever_ access non-readable bytes, that'd be undefined behaviour, might crash the program and even worse things could happen. The non-readable bytes must not be accessible through any (non-unsafe) API, including this one.

Said differently, accessing non-redable bytes is like purposefully reading past the end of a pointer. Something like

```
void *ptr = malloc(10); // get 10 bytes
memset(ptr, 'A', 10); // write 10 As
printf("some unreadable byte: %c"\n", ptr + 100); // read 100 bytes past the pointer (which holds 10 allocated bytes only)
```

Reply:
First off, I'm so grateful for your help and guidance! Coming from the world of web requests and Ruby here, and have a few things to learn. I realize that the whole thing would probably take you under an hour to implement yourself, so I appreciate your time that you spend on coaching me, @weissi. Thank you! 👏🏼 

I'll reply to the conversation threads and try to capture your feedback and work through it.

> @natikgadzhi the readableOnly parameter needs to go. Nothing must ever access non-readable bytes, that'd be undefined behaviour, might crash the program and even worse things could happen. The non-readable bytes must not be accessible through any (non-unsafe) API, including this one.

I take a few things from this: 
- [ ] Remove `readableOnly` as a concept. We will only dump the readable bytes portion.
- [ ] As a rule, only use public API for reading data. Also, dumping is not as complex as to need unsafe access, it feels like. 
- [ ] Agree on what gets dumped, and what is a good public API before building. I want to clarify that: 
   - `hexDump()` should dump readable bytes, which means from `readerIndex` through `writerIndex`. 
   - If we choose to accept `offset` parameter, document that it offsets from the current `readerIndex`, similar to how all `get*(at: Int)` functions currently work.
   - Clarify that it's the user's responsibility to set the `readerIndex` to the desired location before dumping.
   - Clarify that `hexDump` should not move the readerIndex. That is an open questin to me. I _think_ dumping should not change the index (it can just `get*` and that'll do the trick), but I want to confirm that with you. /cc @weissi. 


As for unsafe access for the full buffer allocated length — yep, thank you for the note! I thought that as long as it's within buffer's `capacity`, it would be safe to access and dump, but hey, I'm a newbie here.

Reply:
> I take a few things from this:
> 
> * [ ]  Remove `readableOnly` as a concept. We will only dump the readable bytes portion.
> * [ ]  As a rule, only use public API for reading data. Also, dumping is not as complex as to need unsafe access, it feels like.

That's right

> * [ ]  Agree on what gets dumped, and what is a good public API before building. I want to clarify that:
>   
>   * `hexDump()` should dump readable bytes, which means from `readerIndex` through `writerIndex`.
>   * If we choose to accept `offset` parameter, document that it offsets from the current `readerIndex`, similar to how all `get*(at: Int)` functions currently work.
>   * Clarify that it's the user's responsibility to set the `readerIndex` to the desired location before dumping.
>   * Clarify that `hexDump` should not move the readerIndex. That is an open questin to me. I _think_ dumping should not change the index (it can just `get*` and that'll do the trick), but I want to confirm that with you. /cc @weissi.

Methods for reading from a `ByteBuffer` are usually implemented as `get`/`read` pairs where:
- `getThing(offset: Int) -> Thing?`, reads `Thing` from the given offset or returns `nil` if there weren't enough bytes to read `Thing` or the bytes of `Thing` weren't within the range of the readable bytes. `get` methods don't advance the reader index.
- `readThing() -> Thing?` usually calls `getThing(offset: readerIndex)` and advances the `readerIndex` if `Thing` was read.

As a tool for debugging, we shouldn't be advancing the reader index here. I also don't think we should be able to specify an offset. We should just do a hex dump of all readable bytes in the buffer without advancing the reader index.